### PR TITLE
Return clusterErr rather than err

### PR DESCRIPTION
**Pull request by FengyunPan**
The updateClusterIngressUIDToMasters() should return clusterErr, not err.
If the 'err' of 'masterCluster, masterUID, err := ic.getMasterCluster()' is nil and the 'clusterErr' of 'clusterObj, clusterErr := api.Scheme.DeepCopy(cluster)' is not nil, updateClusterIngressUIDToMasters() will return ("", nil).

And do not log fallbackUID when fallbackUID is nil.

### CODE REVIEW

**Comment by cblecker** [comment]
@k8s-bot ok to test

**Comment by FengyunPan** [comment]
@madhusudancs 
@shashidharatd 
@quinton-hoole 
Can you review this patch? Thank you.

**Comment by FengyunPan** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 1]
```diff
@@ -647,13 +648,15 @@ func (ic *IngressController) getMasterCluster() (master *federationapi.Cluster,
   If there is no master cluster, then fallbackUID is used (and hence this cluster becomes the master).
 */
 func (ic *IngressController) updateClusterIngressUIDToMasters(cluster *federationapi.Cluster, fallbackUID string) (string, error) {
-	masterCluster, masterUID, err := ic.getMasterCluster()
 	clusterObj, clusterErr := api.Scheme.DeepCopy(cluster) // Make a clone so that we don't clobber our input param
 	cluster, ok := clusterObj.(*federationapi.Cluster)
 	if clusterErr != nil || !ok {
-		glog.Errorf("Internal error: Failed clone cluster resource while attempting to add master ingress UID annotation (%q = %q) from master cluster %q to cluster %q, will try again later: %v", uidAnnotationKey, masterUID, masterCluster.Name, cluster.Name, err)
-		return "", err
+		errmsg := fmt.Sprintf("Internal error: Failed clone cluster %q resource: %v", cluster.Name, clusterErr)
```
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: nit: Why 'resource' and not 'error'?  And why remove the detail about UID etc form the log message.  It's useful for debugging, right?

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 2]
```diff
@@ -647,13 +648,15 @@ func (ic *IngressController) getMasterCluster() (master *federationapi.Cluster,
...
+		glog.Errorf(errmsg)
```
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: It's not clear to me that it's a good idea to both log the error and return it (although I realize that the original code did).  Surely the caller will log the error if you return it here?  It seems like the correct general strategy is to either return the error, or log it, not both?

**Review by ghost — COMMENTED** [review]
Thanks @FengyunPan 
Other than a few minor nits this looks good to me.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 1 above)
Comment (FengyunPan) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Yes, you are right, it's useful for debugging. And I am so sorry for my patch's description and removing detail log. I will update it.
I just think the following situation will be bug.
If the 'err' of 'masterCluster, masterUID, err := ic.getMasterCluster()' is nil and the 'clusterErr' of 'clusterObj, clusterErr := api.Scheme.DeepCopy(cluster)' is not nil, updateClusterIngressUIDToMasters() will return ("", nil).

**Comment by FengyunPan** [comment]
@k8s-bot unit test this
@k8s-bot bazel test this

**Comment by quinton-hoole** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44810)*

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 2 above)
Comment (FengyunPan) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: I keep them and just return correct "clusterErr".

**Comment by FengyunPan** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Pull request merged**